### DISCUSSION TITLE
Documentation or Tutorial on Implementing the t-J Model

### DISCUSSION THREAD

**Discussion by coding-physics1** [QuSpin/QuSpin#735] [Q&A]
Dear QuSpin Team :),

I am currently trying to implement the t-J model using QuSpin, as described by the following Hamiltonian:

![Screenshot from 2024-09-17 09-39-00](https://github.com/user-attachments/assets/ebd6be84-99eb-401f-9f44-17a5e0ef1428)

I tried to implement it using the following code:

```from __future__ import print_function, division
import sys, os
import numpy as np
from quspin.basis import spinful_fermion_basis_general
from quspin.operators import hamiltonian


os.environ['KMP_DUPLICATE_LIB_OK'] = 'True'  
os.environ['OMP_NUM_THREADS'] = '1'  
os.environ['MKL_NUM_THREADS'] = '1'  


Lx, Ly = 4, 4  
N_2d = Lx * Ly  
t = 1.0  
J = 0.1  


s = np.arange(N_2d)  
x = s % Lx  
y = s // Lx  
T_x = (x + 1) % Lx + Lx * y  
T_y = x + Lx * ((y + 1) % Ly)  


basis_2d = spinful_fermion_basis_general(N_2d, Nf=(2, 2), double_occupancy=False)


hopping_left = [[-t, i, T_x[i]] for i in range(N_2d)] + [[-t, i, T_y[i]] for i in range(N_2d)]
hopping_right = [[+t, i, T_x[i]] for i in range(N_2d)] + [[+t, i, T_y[i]] for i in range(N_2d)]


J_XX = [[ J, i, T_x[i]] for i in range(N_2d)] + [[ J, i, T_y[i]] for i in range(N_2d)]
J_YY = [[J, i, T_x[i]] for i in range(N_2d)] + [[ J, i, T_y[i]] for i in range(N_2d)]
J_ZZ = [[J, i, T_x[i]] for i in range(N_2d)] + [[J, i, T_y[i]] for i in range(N_2d)]


static = [["+-|", hopping_left],
          ["-+|", hopping_right],
          ["|+-", hopping_left],
          ["|-+", hopping_right],
          ["xx|", J_XX],
          ["yy|", J_YY],
          ["zz|", J_ZZ]]


H = hamiltonian(static, [], basis=basis_2d, dtype=np.complex128)


E_GS, psi_GS = H.eigsh(k=4, which='SA')  


print('E_GS', E_GS)
```

However, when comparing my results to other exact diagonalization (ED) codes, I get different results for the ground state energy. I'm wondering if I'm implementing something incorrectly or missing a step in setting up the model.

Is there any available documentation or tutorial that explains how to set up and implement the t-J model correctly in QuSpin? Any guidance or pointers would be greatly appreciated!

Thank you for your help! :) 

I had a look in

https://github.com/QuSpin/QuSpin/blob/master/test/test_sq_lat_Heis_double_occupancy.py

kindly pointed out by mgbukov :), but I am not sure if this corresponds to the model I am trying to compute, as I don't want double occupancy...

**Comment by mgbukov** [comment]
difficult to say right away; why don't you take a look at our comparison here, maybe u spot what's going wrong:

https://github.com/QuSpin/QuSpin/blob/master/test/test_sq_lat_Heis_double_occupancy.py

**Reply by coding-physics1** [reply]
thank you for your help, I will have another look and keep you updated
